contracts/cw1155-base/src/contract.rs

```diff
@@ -0,0 +1,1239 @@
+            msg,
+        } => execute_batch_send_from(env, from, to, batch, msg),
+        Cw1155ExecuteMsg::Mint {
+            to,
+            token_id,
+            value,
+            msg,
+        } => execute_mint(env, to, token_id, value, msg),
+        Cw1155ExecuteMsg::BatchMint { to, batch, msg } => execute_batch_mint(env, to, batch, msg),
+        Cw1155ExecuteMsg::Burn {
+            from,
+            token_id,
+            value,
+        } => execute_burn(env, from, token_id, value),
+        Cw1155ExecuteMsg::BatchBurn { from, batch } => execute_batch_burn(env, from, batch),
+        Cw1155ExecuteMsg::ApproveAll { operator, expires } => {
+            execute_approve_all(env, operator, expires)
+        }
+        Cw1155ExecuteMsg::RevokeAll { operator } => execute_revoke_all(env, operator),
+    }
+}
+
+/// When from is None: mint new coins
+/// When to is None: burn coins
+/// When both are None: not token balance is changed, meaningless but valid
+fn execute_transfer_inner<'a>(
+    deps: &'a mut DepsMut,
+    from: Option<&'a HumanAddr>,
+    to: Option<&'a HumanAddr>,
+    token_id: &'a str,
+    amount: Uint128,
+) -> Result<TransferEvent<'a>, ContractError> {
+    if let Some(from) = from {
+        let from_raw = deps.api.canonical_address(from)?;
+        BALANCES.update(
+            deps.storage,
+            (from_raw.as_slice(), token_id),
+            |balance: Option<Uint128>| balance.unwrap_or_default() - amount,
+        )?;
+    }
+
+    if let Some(to) = to {
+        let canonical_to = deps.api.canonical_address(to)?;
+        BALANCES.update(
+            deps.storage,
+            (canonical_to.as_slice(), token_id),
+            |balance: Option<Uint128>| -> StdResult<_> { Ok(balance.unwrap_or_default() + amount) },
+        )?;
+    }
+
+    Ok(TransferEvent {
+        from,
+        to,
+        token_id,
+        amount,
+    })
+}
+
+/// returns true iff the sender can execute approve or reject on the contract
+fn check_can_approve(
+    deps: Deps,
+    env: &Env,
+    owner: &HumanAddr,
+    operator: &HumanAddr,
+) -> StdResult<bool> {
+    // owner can approve
+    let owner_raw = deps.api.canonical_address(owner)?;
+    let operator_raw = deps.api.canonical_address(operator)?;
+    if owner_raw == operator_raw {
+        return Ok(true);
+    }
+    // operator can approve
+    let op = APPROVES.may_load(deps.storage, (&owner_raw, &operator_raw))?;
+    Ok(match op {
+        Some(ex) => !ex.is_expired(&env.block),
+        None => false,
+    })
+}
+
+fn guard(cond: bool, err: ContractError) -> Result<(), ContractError> {
+    if cond {
+        Ok(())
+    } else {
+        Err(err)
+    }
+}
+
+pub fn execute_send_from(
+    env: ExecuteEnv,
+    from: HumanAddr,
+    to: HumanAddr,
+    token_id: TokenId,
+    amount: Uint128,
+    msg: Option<Binary>,
+) -> Result<Response, ContractError> {
+    let ExecuteEnv {
+        mut deps,
+        env,
+        info,
+    } = env;
+
+    guard(
```

Comment:
You find this easier to write than

```rust
if !check_can_approve(deps.as_ref(), &env, &from, &info.sender)? {
  return Err(ContractError::Unauthorized {});
}
```

?? It's fine either way.

Reply:
agree, it's pointless, removed the `guard`, ;D

Reply:
Further encapsulate a `guard_can_approve` which return `ContractError` directly.

Reply:
> Further encapsulate a guard_can_approve which return ContractError directly.

Yes, something like `guard_can_approve(deps.as_ref(), &env, &from, &info.sender)?;` or `assert_can_approve(deps.as_ref(), &env, &from, &info.sender)?;` seems like a nice abstraction, just not the two separately which doesn't save typing